src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelLockFileModule.java

```diff
@@ -64,26 +65,37 @@ public void afterCommand() throws AbruptExitException {
     if (moduleResolutionEvent == null) {
       // Command does not use Bazel modules or the lockfile mode is not update.
       // Since Skyframe caches events, they are replayed even when nothing has changed.
-      Preconditions.checkState(extensionResolutionEventsMap.isEmpty());
       return;
     }
 
+    // All nodes corresponding to module extensions that have been evaluated in the current build
+    // are done at this point. Look up entries by unvalidated keys to record results even if
+    // validation later fails due to invalid imports.
+    // Note: This also picks up up-to-date result from previous builds that are not in the
+    // transitive closure of the current build. Since extension are potentially costly to evaluate,
+    // this is seen as an advantage. Full reproducibility can be ensured by running 'bazel shutdown'
+    // first if needed.
+    Map<ModuleExtensionId, LockFileModuleExtension.WithFactors> newExtensionInfos =
+        new ConcurrentHashMap<>();
+    evaluator
+        .getInMemoryGraph()
+        .parallelForEach(
```

Comment:
I found this to be much faster (8x on a slow machine, 1.5x on a fast machine) than going through `getDoneValues`, even with a `parallelStream`, on my machine. It's also not much more complicated.